src/main/java/space/obminyashka/items_exchange/service/impl/AdvertisementServiceImpl.java

```diff
@@ -63,6 +65,20 @@ public List<AdvertisementTitleDto> findRandom12Thumbnails() {
         return findAllThumbnails(PageRequest.of(randomPage, resultsQuantity)).getContent();
     }
 
+    @Override
+    @Cacheable
+    public List<AdvertisementTitleDto> findRandom4AdvertisementWithSameSubcategory(UUID advertisementId,
+                                                                                   Long subcategoryId,
+                                                                                   Pageable pageable) {
+        final var totalRecordsSize = advertisementRepository.countByIdNotAndSubcategoryId(advertisementId, subcategoryId);
+        final var bound = (int) (totalRecordsSize / pageable.getPageSize());
+        final var randomPageNumber = bound > 0 ? random.nextInt(bound) : 0;
+        return advertisementRepository.findAllByIdNotAndSubcategoryId(advertisementId, subcategoryId,
+                        PageRequest.of(randomPageNumber, pageable.getPageSize()))
```

Comment:
You receive incoming **page** and **size** parameters as incoming parameters in the endpoint but doing some hiding operations with them here.